javascript/javascript3/week3/ShoppingCart/ShopCartJS.js

```diff
@@ -0,0 +1,87 @@
+const productContainer = document.getElementById("products-container");
+
+class Product {
+  constructor(name, price) {
+    this.name = name;
+    this.price = price;
+  }
+}
+
+class ShoppingCart {
+  constructor() {
+    this.products = [];
+  }
+
+  addProduct(product) {
+    this.products.push(product);
+    this.currency = product.currency;
+  }
+
+  removeProduct(product) {
+    const index = this.products.indexOf(product);
+    if (index > -1) {
+      this.products.splice(index, 1);
+    }
+  }
+
+  searchProduct(productName) {
+    const searchedProduct = this.products.filter((product) =>
+      product.name.toLowerCase().includes(productName)
+    );
+    return console.log(searchedProduct);
+  }
+
+  renderProducts() {
+    this.products.forEach((product) => {
+      const productList = document.createElement("p");
+      productList.innerHTML = `<p>${product.name}: ${product.price}</p>`;
+      productContainer.appendChild(productList);
+    });
+  }
+
+  getTotal() {
+    const allProducts = this.products;
+    const totalPrice = allProducts
+      .map((productPrice) => productPrice.price)
+      .reduce(
+        (previousValue, nextProductValue) => previousValue + nextProductValue,
+        0
+      );
+    const totalPricePara = document.createElement("p");
+    totalPricePara.innerText = `Total price of shopping cart: ${totalPrice}`;
+    productContainer.appendChild(totalPricePara);
+  }
+
+  getUser() {
+    const randomUser = Math.floor(Math.random() * 10 + 1);
+    return fetch(
+      `https://jsonplaceholder.typicode.com/users/${randomUser}`
+    ).then((response) =>
+      response.json().then((data) => {
+        const userNameContainer = document.getElementById("username-container");
+        const userName = document.createElement("p");
+        userName.innerHTML = data.name;
+        userNameContainer.appendChild(userName);
+      })
```

Comment:
We're falling back into the nested callback hell trap like this by making a promise chain inside another. We can leverage the `.then` call to have less nesting:
```
.then((response) => response.json())
.then((data) => {
    ....
}
```
or by using `await` :
```
.then((response) => {
    const data = await response.json()
    ....
}
```